TrueDialog.Client.Standard/Client/Context/BaseContext.cs

```diff
@@ -0,0 +1,43 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Reflection;
+
+namespace TrueDialog.Context
+{
+    internal class BaseContext
+    {
+        protected readonly IApiCaller Api;
+
+        public BaseContext(IApiCaller api)
+        {
+            Api = api;
+        }
+
+        protected int CurrentAccount { get { return Api.AccountId; } }
+
+        #region SubContext factory
+        protected readonly ConcurrentDictionary<string, BaseContext> m_contextDictionary = new ConcurrentDictionary<string, BaseContext>();
+        private readonly object m_mutex = new object();
```

Comment:
Replace with SemaphoreSlim

Reply:
For that matter, it might be possible to do away with this all together depending on how carefully we use the m_contextDictionary because it's Concurrent

Reply:
Scratch that, we don't need any of it.  We can replace this with DI.

Reply:
How can we use DI inside of a class library? Well, in some cases we can (the same way as we expose services.AddTrueDialog()), but that won't work in all use cases.

Reply:
How you normally would use DI.
```csharp
public class Widget
{
    private readonly IAccountContext m_accountContext;

    public Widget(IAccountContext accountContext)
    {
        m_accountContext = accountContext;
    }

    public void DoThing()
    {
        // Do something with m_accountContext
    }
}
```

Then you just register the account context with DI:
```csharp
container.RegisterScoped<IAccountContext, AccountContext>();
```

The class library just needs a reference to the Microsoft.DI.Abstractions library.